tutorials/synthetic_retrieval_evaluation_customization/retriever_evalset_generator.py

```diff
@@ -0,0 +1,191 @@
+    """
+
+    def __init__(
+        self,
+        pipeline_config: RetrieverEvalSDGConfig = None,
+    ):
+        super().__init__()
+        self._name = self.__class__.__name__
+        self.cfg = pipeline_config
+
+        self._init_pipeline_params()
+
+    def load_pipeline_config(self, cfg_path: str):
+        self.cfg = RetrieverEvalSDGConfig.from_yaml(cfg_path)
+
+    def _validate_config(self):
+        return True  # TODO complete this
+
+    def _init_pipeline_params(self):
+        # synchronous
+        self.openai_client = OpenAI(
+            base_url=self.cfg.base_url,
+            api_key=self.cfg.api_key,
+        )
+        self.client = OpenAIClient(self.openai_client)
+        self.generator = NemotronGenerator(self.client)
+
+        # TODO asynchronous
+        self.async_openai_client = AsyncOpenAI(
+            base_url=self.cfg.base_url,
+            api_key=self.cfg.api_key,
+        )
+        self.async_client = AsyncOpenAIClient(self.async_openai_client)
+        self.async_generator = AsyncNemotronGenerator(self.async_client)
+
+        if self._validate_config():
+            self.sys_prompt = self.cfg.generator_system_prompt
+            self.user_prompt_template = self.cfg.generator_user_prompt_template
+            self.generator_model = self.cfg.generator_model
+            self.generator_model_kwargs = {
+                "temperature": self.cfg.temperature,
+                "top_p": self.cfg.top_p,
+                "max_tokens": self.cfg.max_tokens,
+            }
+            self.num_qs = self.cfg.num_questions
+        else:
+            raise Exception("Validation Error: incorrect pipeline config file")
+
+    # ----------------------------------------------------------------------------80
+
+    def __call__(self, dataset: DocumentDataset) -> DocumentDataset:
+
+        df = dataset.df
+
+        df["llm_response"] = df["text"].apply(
+            self.generate, meta=("llm_response", "str")
+        )
+        df["qa_pairs"] = df["llm_response"].apply(
+            self.parse_response, meta=("qa_pairs", "object")
+        )
+
+        df = df.explode("qa_pairs").reset_index(drop=True)
+
+        df["question"] = df["qa_pairs"].apply(
+            lambda x: x["question"], meta=("question", "str")
+        )
+
+        if "_id" in df.columns:
+            df["_id"] = df["_id"].apply(self._check_doc_id, meta=("_id", "str"))
+        else:
+            df["_id"] = df["text"].apply(self._get_random_hash, meta=("_id", "str"))
+
+        df["question-id"] = df["question"].apply(
+            self._get_random_hash, meta=("question-id", "str")
+        )
+
+        df["answer"] = df["qa_pairs"].apply(
+            lambda x: x["answer"], meta=("answer", "str")
+        )
+
+        df["score"] = df["question"].apply(lambda x: 1, meta=("score", "int"))
+
+        return DocumentDataset(
+            df[["_id", "text", "title", "question-id", "question", "answer", "score"]]
```

Comment:
Instead of selecting what fields to preserve, can you use `df.drop` to remove any unwanted columns? The user may have more fields in their dataset that you are unaware of.

Reply:
fixed, thanks!